src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,309 @@
+- Signature bytes (based on dispatch)
+
+Lease and key lease fields (when not elided) are encoded using the compact integer format.
+
+## Decoding Rules
+
+Decoding rules govern the reconstruction of the SRP message. They dictate the sequence in which resource records are added and the proper usage of DNS name compression (pointer names). This guarantees an exact replication of the original SRP message, a critical step for validating the calculated signature, which is based on the original message.
+
+Domain name (e.g., "default.service.arpa") is included in the SRP update message within the Zone section. Subsequent references to this domain name within other DNS names MUST use pointers (DNS name compression).
+
+The first occurrence of the host name (e.g., in an SRV record or an AAAA record) directly appends its labels (preceding the domain name). For any subsequent appearances of the host name (in SRV, AAAA, KEY records), pointer names (name compression) MUST be employed to refer to its first appearance.
+
+#### An "Add Service Block" is decoded as follows:
+
+1. A PTR record mapping the service name to the service instance.
+2. PTR records for any sub-types (if present).
+3. A "Delete all RRsets from a name" record for the service instance name.
+4. An SRV record for the service instance.
+5. A TXT record for the service instance.
+
+- The service type labels (e.g., `_matter._tcp`) in the first PTR record MUST be appended directly, even if the same service name appears earlier in the message. Pointer names (compressed names) MUST NOT be used in this case.
+- The service instance name in the first PTR record appends the instance label directly, followed by a pointer to the service type name from the PTR record name. After the first PTR record, a pointer to this name MUST be used to refer to the service instance name.
+- The first sub-type PTR appends the sub-type label, followed by the `_sub` label with a pointer to the service name. Any subsequent sub-type appends the sub-type label, followed by a pointer to the `._sub` name from the first sub-type PTR.
+- If TXT data is elided, a TXT record is still appended with a single zero byte (empty TXT data).
+
+#### A “Remove Service block” is decoded as:
+
+1. A PTR as “Delete an RR from an RRSet" where class is set to NONE and TTL to zero (RFC 2136 - section 2.5.4)
+
+#### The “Host Block” is decoded as follows:
+
+1. A “Delete all RRsets from a name” for the host name.
+2. For each address in order they appear in the encoded message, an AAAA record is appended.
+3. If key is provided, a KEY record is appended for the host name
+
+#### The “Footer Block” is decoded as follows:
+
+1. An OPT record is appended with single Lease Option (in Additional Data section)
+   - Empty (root domain) is used for OPT RR name
+   - The long variant of Lease Option is used setting lease and key lease intervals
+2. If signature is provided, a SIG record is appended following SRP protocol specification.
```

Comment:
I don't see this talking at all about how the signature is generated. Presumably we take a signed uncompressed packet, compress it, and then uncompress it, but the compression format constrains the ordering of SRP info in the compressed message in ways that aren't requires by the SRP spec, so there are valid SRP messages for which the ordering on the way in would be different than on the way out.

Also, keys are optional for instances, but we don't have a way to say whether or not the key was present for an instance.

I would suggest that the way to fix this is to require that data in the SRP message be sent in an order that is consistent with how it would show up in the compressed message, and to require that the SRP message used as input for the compressor be in the correct order; if it is not, the compressor would have to reject it (that is, not compress it).

Reply:
Good point. Added more about this at the beginning of the "Decoding Rules" section.

> I would suggest that the way to fix this is to require that data in the SRP message be sent in an order that is consistent with how it would show up in the compressed message, and to require that the SRP message used as input for the compressor be in the correct order; if it is not, the compressor would have to reject it (that is, not compress it).

Yes, exactly. That's the idea.